src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -206,13 +236,58 @@ private void Stop()
                 IPEndPoint remoteEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref *(SOCKADDR_INET*)connectionInfo.RemoteAddress);
+                string targetHost = string.Empty;   // compat with SslStream
+                if (connectionInfo.ServerNameLength > 0 && connectionInfo.ServerName != IntPtr.Zero)
+                {
+                    // TBD We should figure out what to do with international names.
+                    targetHost = Marshal.PtrToStringAnsi(connectionInfo.ServerName, connectionInfo.ServerNameLength);
+                    //Console.WriteLine("Listener got connected to {0}", targetHost);
+                }
+
+                SafeMsQuicConfigurationHandle? connectionConfiguration = state.ConnectionConfiguration;
+
+                if (state.ServerOptionsSelectionCallback != null)
+                {
+                    // SslClientHelloInfo does not have public constructor (yet?)
+                    object[] parameters = new object[2];
+                    parameters[0] = targetHost;
+                    parameters[1] = SslProtocols.Tls13;
+                    SslClientHelloInfo info = (SslClientHelloInfo)_SslClientHelloInfo.Invoke(parameters);
+
+                    // We should figure out what to do with cancellation.
+                    ValueTask<SslServerAuthenticationOptions> t = state.ServerOptionsSelectionCallback(state, info, state.ServerOptionsSelectionCallbackState, default);
+                    if (t.IsCompletedSuccessfully)
+                    {
+                        // Easy pass, we completed synchronously. We can take similar path to ServerCertificateSelectionCallback.
+                        connectionConfiguration = SafeMsQuicConfigurationHandle.Create(state.ConnectionOptions, t.Result, targetHost);
+                    }
+                    else
+                    {
+                        // We can return MsQuicStatusCodes.Pending and park the processing.
+                        // MsQuic should wait either for setting configuration or for closing the connection.
+                        // For now we will block and the result.
+
+                        SslServerAuthenticationOptions options = t.AsTask().GetAwaiter().GetResult();
+                        connectionConfiguration = SafeMsQuicConfigurationHandle.Create(state.ConnectionOptions, options, targetHost);
+                    }
+                }
+                else if (state.AuthenticationOptions.ServerCertificateSelectionCallback != null)
+                {
+                    // ServerCertificateSelectionCallback is synchronous. We will call it as needed when building configuration
+                    connectionConfiguration = SafeMsQuicConfigurationHandle.Create(state.ConnectionOptions, state.AuthenticationOptions, targetHost);
+                }
+
+                if (connectionConfiguration == null)
+                {
+                    return MsQuicStatusCodes.InternalError;
```

Comment:
This seems bad. When would this happen?

Reply:
This is same as all the other case when for example the callback throw -> we would abort the Listener. 
It feels the handling `NEW_CONNECTION` should not abort the listener but I did not want to add that to SNI change.
Perhaps this is different as we running user code and processing user data. 
But let say client sends request for SNI server is not configured to handle (or does not want to for what ever reason) That should not bring down whole server. Or the callback fails to connect to database. 
Perhaps we explicitly close the handle and return OK to MsQuic? cc: @nibanks

Reply:
I'm unclear on exactly what the question is here. With any NEW_CONNECTION event you have the following options:

1. Return a failure, and MsQuic will take care of cleaning up the connection.
2. Return success/pending (we don't actually differentiate) and the app now owns the connection.
  a. You can set the configuration inline, and MsQuic will immediate continue execution of the handshake.
  b. You can set the configuration on a separate thread, which will cause MsQuic to then continue the handshake.
  c. You can shutdown the connection.

Reply:
Based on my testing we loose the whole listener. I want to make sure that what ever we do impacts only the new connection and not the listener new connection was received on. It _is_ possible this is caused by some our code - I would need to dig deeper.

Reply:
I read what Nick said as meaning, if we don't want to accept this connection, we should shut it down in the callback here. (As opposed to returning InternalError as we do above.)

Reply:
yes, I agree. I'll make the change.

Reply:
I end up leaving the InternalError there but reworked the cleanup logic. Here we don't have SafeHandle yet so leaving  MsQuic to clean it up immediately seems like the most straight-forward approach. 
I modified the tests to make sure we can fail one connection but the listener stays functional.